### PR TITLE
Adds cache option to uplinks

### PR DESCRIPTION
**Pull request by silkentrance**
Open to discussion. Please review.

*WIP*: needs tests

[GH-131] add cache option to uplinks

### CODE REVIEW

File: lib/storage.js [hunk 1]
```diff
@@ -246,7 +246,11 @@ Storage.prototype.get_tarball = function(name, filename) {
       }, self.config)
     }
 
-    var savestream = self.local.add_tarball(name, filename)
+    var savestream
+    var docache = !!uplink.cache
```
Comment (jachstet-sea) [inline lib/storage.js]: Just curious: Any reason to have a double negation here?

Comment (silkentrance) [inline lib/storage.js]: Yes, you are correct, this is rather unneccessary. Out of habit, I think...

**Comment by juanpicado** [comment]
I'm considering to merge this one, but I need a rebase from master and unit testing. Could you take a look #142 @silkentrance  ?

**Comment by silkentrance** [comment]
@juanpicado Will do.

**Comment by juanpicado** [comment]
@silkentrance sure, no problem, take your time 👍

**Comment by silkentrance** [comment]
I also fixed the existing integration tests, namely I adjusted the sinopia-test-1.2.3.tgz package to match the requirements, renaming everything to verdaccio.

I also added test/integration/config_nocache.yaml and test/integration/test_nocache.pl.

I think that these integration tests need to be re-implemented :grin:

The configuration uplink caching behavior is tested in test/unit/config.js.

**Comment by juanpicado** [comment]
Awesome, I'll review soon 👍

**Comment by silkentrance** [comment]
ATM I am working on implementing the functional tests, but this may take a while. I need to get my head around the test cases first.

**Comment by silkentrance** [comment]
@juanpicado I am having a hard time with the proper configuration of server3 (config-3.yaml) in test/functional. It will always fail with a 404 on get_package.

While the integration test works as expected, I cannot get it to work in the functional tests.

Maybe you can give me some hints on how to use the functional server test API?

File: lib/storage.js [hunk 2]
```diff
@@ -278,10 +278,14 @@ class Storage {
       if (uplink == null) {
         uplink = new Proxy({
           url: file.url,
+          cache: true,
           _autogenerated: true,
         }, self.config);
       }
-      let savestream = self.local.add_tarball(name, filename);
+      let savestream = null;
+      if (uplink.config.cache == true) {
```
Comment (juanpicado) [inline lib/storage.js]: Do we need `==` for true?

File: test/functional/gh131.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+const path = require('path');
+const assert = require('assert');
+const crypto = require('crypto');
+
+const STORAGE = 'test-storage3';
+const TARBALL = 'blahblah';
+const PKG_NOCACHE = 'pkg_nocache';
+const PKG_NOCACHE2 = 'pkg_nocache2';
+
+function iscached(pkgname, tarballname) {
```
Comment (juanpicado) [inline test/functional/gh131.js]: camelCase here please, `isCache`.  The same in other functions. I'll enable camelCase eslint rule soon.

File: test/functional/gh131.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  });
+
+  before(function() {
+    return server2.put_tarball(PKG_NOCACHE2, TARBALL, readfile('fixtures/binary'))
+             .status(201)
+             .body_ok(/.*/);
+  });
+
+  before(function() {
+    let pkg = require('./lib/package')(PKG_NOCACHE2);
```
Comment (juanpicado) [inline test/functional/gh131.js]: const if the variable does not change

File: test/functional/lib/startup.js [hunk 5]
```diff
@@ -41,4 +42,5 @@ module.exports.start = function(dir, conf) {
 process.on('exit', function() {
   if (forks[0]) forks[0].kill();
```
Comment (juanpicado) [inline test/functional/lib/startup.js]: Let's use `curly` braces, I'll enable that rule soon as well.

File: test/unit/config.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+let assert = require('assert');
```
Comment (juanpicado) [inline test/unit/config.js]: `const` for imports

File: test/unit/partials/config.js [hunk 7]
```diff
@@ -2,6 +2,11 @@
 
 let config = {
```
Comment (juanpicado) [inline test/unit/partials/config.js]: `const` here as well

**Comment by juanpicado** [comment]
@silkentrance I'm testing your branch locally, to figure out the `404`.

**Comment by silkentrance** [comment]
@juanpicado will fix. thanks for looking into this. :+1:

**Comment by juanpicado** [comment]
@silkentrance the `404` seems to be because in your `packages` configuration for server3 you define `pkg-nocache` while in your test you define

```javascript
const PKG_NOCACHE = 'pkg_nocache';
const PKG_NOCACHE2 = 'pkg_nocache2';
```
Thus, the uplinks are false, that drives you to 404. Then if you fix that the asserts are false, I do not know if the `PKG_NOCACHE` name is intended or not.

**Comment by silkentrance** [comment]
@juanpicado Thanks so much for Findling

**Comment by silkentrance** [comment]
@juanpicado Thank you so much for finding this annoying typo. I will fix that right awy incl. the issues you found in your review.

**Comment by silkentrance** [comment]
Dunno why this was closed. Must have hit the wrong button...

**Comment by silkentrance** [comment]
@juanpicado fixed issues and tests. rebased to master.

**Comment by juanpicado** [comment]
Cool !! @silkentrance I'll review it asap

**Comment by juanpicado** [comment]
Awesome work, unit testing widely complete 👍  thanks so much ! This goes to the next version.

**Pull request merged**